aten/src/ATen/core/dispatch/DispatchKeyExtractor.h

```diff
@@ -48,13 +48,87 @@ namespace detail {
     }
   };
 
+  struct TensorOptionsAccumulator : at::IterArgs<TensorOptionsAccumulator> {
+    TensorOptions options;
+    void operator()(c10::optional<ScalarType> dtype) {
+      if (dtype.has_value()) {
+        options = options.dtype(*dtype);
+      }
+      else {
+        // Following the TensorOptions logic where in case of undefined
+        // dtype, default dtype is returned.
+        // Issue https://github.com/pytorch/pytorch/issues/30405
+        options = options.dtype(at::get_default_dtype());
+      }
+    }
+    void operator()(c10::optional<Device> device) {
+      if (device.has_value()) {
+        options = options.device(*device);
+      }
+    }
+    void operator()(c10::optional<Layout> layout) {
+      if (layout.has_value()) {
+        options = options.layout(*layout);
+      }
+    }
+    void operator()(c10::optional<bool> pin_memory) {
+      if (pin_memory.has_value()) {
+        options = options.pinned_memory(*pin_memory);
+      }
+    }
+    void operator()(ScalarType dtype) {
+      options = options.dtype(dtype);
+    }
+    void operator()(Device device) {
+      options = options.device(device);
+    }
+    void operator()(Layout layout) {
+      options = options.layout(layout);
+    }
+    void operator()(bool pin_memory) {
```

Comment:
this is dangerous. Even if the operator has the correct set and order of TensorOptions args, there might be additional bool arguments and they will go into here even though they're not `pin_memory`. You need to make sure that only the `bool` argument that comes directly after the `Layout` argument goes into here. This probably means that you cannot use `IterArgs` anymore but need to write your own metaprogramming.